### ISSUE TITLE
Utilization for Enclosure and Servers

### ISSUE THREAD

**Issue by hershsa** [HewlettPackard/POSH-HPOneView#87]
Feature request to include utilization for Enclosures and servers - /rest/server-hardware/{id}/utilization

Please fill in as much information as possible to help resolve your issue.

### Expected Behavior

utilization is not returned with the get-hpovserver command
### Actual Behavior
property is not available.

### Version Information
HPE OneView PowerShell Library Version (`Get-HPOVVersion`): 2.0.300.0 
HPE OneView Appliance Version:   2.00.02-0219211
Output from `$PSVersionTable` on your Windows Host:
Name                           Value                                                                                                                                                          
----                           -----                                                                                                                                                          
PSVersion                      5.0.10240.16384                                                                                                                                                
WSManStackVersion              3.0                                                                                                                                                            
SerializationVersion           1.1.0.1                                                                                                                                                        
CLRVersion                     4.0.30319.42000                                                                                                                                                
BuildVersion                   10.0.10240.16384                                                                                                                                               
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}                                                                                                                                        
PSRemotingProtocolVersion      2.3

**Comment by ChrisLynchHPE** [comment]
We will look at adding this in a future release.

Would you really want utilization data (Power, Thermal and CPU) to be reported with the `Get-HPOVServer` Cmdlet, or it's own?  I ask because utilization data is not part of the server resource object via the API.  It is a completely different API call.

**Comment by hershsa** [comment]
Ahh looks like I misread the documentation. On its own probably makes the most sense then. Guessing the thought would be you would pass either a server/enclosure object to it to retrieve its utilization data?

**Comment by ChrisLynchHPE** [comment]
> Guessing the thought would be you would pass either a server/enclosure object to it to retrieve its utilization data?

That would be likely the idea.  Similar to how you can get Interconnect utilization data with the `Show-HPOVPortStatistic` Cmdlet, you can pipe an Interconnect object to it.

**Comment by hershsa** [comment]
I noticed in another issue post you recommended using Send-HPOVRequest, would that cmdlet work for me in the meantime as well? Something like this:

$serverUtil = Send-HPOVRequest ($uri = '/rest/server-hardware/{id}/utilization')

**Comment by ChrisLynchHPE** [comment]
Yep.  `Send-HPOVRequest` is the core Cmdlet that all other Cmdlets use to interact with an appliance API.  Although, the command would be:

```PowerShell
$Server = Get-HPOVServer 'Enc1, Bay 1'
$ServerUtilization = Send-HPOVRequest ($Server.uri + '/utilization')
```

**Comment by hershsa** [comment]
Thanks for the quick responses. Will give this a try tomorrow.

**Comment by hershsa** [comment]
This worked like a charm, thanks!

**Comment by ChrisLynchHPE** [comment]
Glad it did.  I'm still going to keep this open, and will add as an enhancement to a future release.

**Comment by ChrisLynchHPE** [comment]
@hershsa have you been able to get the data you were looking for?  I'm looking for input or opinion on the Cmdlet for this functionality.  Would you expect the Cmdlet to return the raw utilization data, or provide custom values back?  For instance, the `metricSample` value(s) are based on millisecond passed since Epoch time (01/01/1970 12:00).  I'm think that the Cmdlet would just provide summary data for CPU, RAM and Thermal data for a supported Server resource.  Thoughts?

**Comment by hershsa** [comment]
Yes I have been. I basically took the metricSample values and calculated my average from all the entries and found the peak in that same time frame as well.

I would lean towards custom values. So maybe average and peak over the past 24 hours or the entire data sample. I think that is more likely to be used then the raw data.

**Comment by ChrisLynchHPE** [comment]
Okay. So how about an PSCustomObject with the following:

| CPUAvg | CPUPeak | PowerAvg | PowerPeak | ThermalAvg | ThermalPeak |
---------------|----------------|----------------|-----------------|-------------------|--------------------|
| 12 | 50 | 50 | 300 | 75f | 82F |

**Comment by hershsa** [comment]
That looks great.

**Comment by ChrisLynchHPE** [comment]
Okay.  I''ll add this to the next release I'll publish this week.

**Comment by freedge** [comment]
Has the Show-HPOVUtilization ever made it into the HPOneview module? I am using HPOneView.500 version 5.0.2400.1845 and still will have to use the workaround with Send-HPOVRequest

**Comment by ChrisLynchHPE** [comment]
Apologies.  Yes, this is bring wrapped up now.  I realize this is extremely late, but this got away from my radar.  This is what the Cmdlet will look like:

```powershell
# Show utilization output using default format view
[PS] C:\> Get-OVServer -Name 'TBIRD6-E1, bay 7' | Show-OVUtilization

Name             CPUCurrent PowerCurrent AmbientTemperature LastUpdated
----             ---------- ------------ ------------------ -----------
TBIRD6-E1, bay 7 82         61 W         79 °F              10/1/2020 10:45:00 PM

# Display utilization object via Format-List to see all properties
[PS] C:\> Get-OVServer -Name 'TBIRD6-E1, bay 7' | Show-OVUtilization | Format-List

Name                      : TBIRD6-E1, bay 7
Uri                       : /rest/server-hardware/39313738-3634-584D-5138-313730365A31
CPUCurrent                : 82
CPUAverage                : 82
CPUPeak                   : 82
PowerCurrent              : 61 W
PowerAverage              : 62 W
PowerPeak                 : 63 W
AmbientTemperature        : 79 °F
AmbientTemperatureAverage : 79 °F
LastUpdated               : 10/1/2020 10:45:00 PM
PollingInterval           : 00:05:00
ApplianceConnection       : 16.83.5.226

[PS] C:\> Get-OVEnclosure | Show-OVUtilization

Name      PowerCurrent AmbientTemperature LastUpdated
----      ------------ ------------------ -----------
TBIRD6-E1 1,151 W      84 °F              10/1/2020 10:45:00 PM
TBIRD6-E2 927 W        84 °F              10/1/2020 10:45:00 PM
TBIRD6-E3 674 W        84 °F              10/1/2020 10:50:00 PM

# Display server utilization object properties and methods from Get-Member
[PS] C:\> Get-OVServer -Name 'TBIRD6-E1, bay 7' | Show-OVUtilization | Get-Member -view all

   TypeName: HPEOneView.ServerUtilization

Name                      MemberType Definition
----                      ---------- ----------
Equals                    Method     bool Equals(System.Object obj)
GetHashCode               Method     int GetHashCode()
GetType                   Method     type GetType()
ToString                  Method     string ToString()
AmbientTemperature        Property   HPEOneView.Temperature AmbientTemperature {get;set;}
AmbientTemperatureAverage Property   HPEOneView.Temperature AmbientTemperatureAverage {get;set;}
ApplianceConnection       Property   HPEOneView.Library.ApplianceConnection ApplianceConnection {get;set;}
CPUAverage                Property   int CPUAverage {get;set;}
CPUCurrent                Property   int CPUCurrent {get;set;}
CPUPeak                   Property   int CPUPeak {get;set;}
LastUpdated               Property   datetime LastUpdated {get;set;}
Name                      Property   string Name {get;set;}
PollingInterval           Property   timespan PollingInterval {get;set;}
PowerAverage              Property   HPEOneView.Power PowerAverage {get;set;}
PowerCurrent              Property   HPEOneView.Power PowerCurrent {get;set;}
PowerPeak                 Property   HPEOneView.Power PowerPeak {get;set;}
Uri                       Property   string Uri {get;set;}

# Show enclosure utilization object properties and methods from Get-Member
[PS] C:\> Get-OVEnclosure -Name TBIRD6-E1 | Show-OVUtilization | Get-Member -view all

   TypeName: HPEOneView.EnclosureUtilization

Name                      MemberType Definition
----                      ---------- ----------
Equals                    Method     bool Equals(System.Object obj)
GetHashCode               Method     int GetHashCode()
GetType                   Method     type GetType()
ToString                  Method     string ToString()
AmbientTemperature        Property   HPEOneView.Temperature AmbientTemperature {get;set;}
AmbientTemperatureAverage Property   HPEOneView.Temperature AmbientTemperatureAverage {get;set;}
ApplianceConnection       Property   HPEOneView.Library.ApplianceConnection ApplianceConnection {get;set;}
LastUpdated               Property   datetime LastUpdated {get;set;}
Name                      Property   string Name {get;set;}
PollingInterval           Property   timespan PollingInterval {get;set;}
PowerAverage              Property   HPEOneView.Power PowerAverage {get;set;}
PowerCurrent              Property   HPEOneView.Power PowerCurrent {get;set;}
PowerPeak                 Property   HPEOneView.Power PowerPeak {get;set;}
Uri                       Property   string Uri {get;set;}
```

**Comment by davidvf** [comment]
Could we also get support for "Get-OVInterconnect" as a valid input object?

**Comment by ChrisLynchHPE** [comment]
What specifically are you looking for?  If interconnect port stats, that's what [Show-OVPortStatistics](https://hpe-docs.gitbook.io/posh-hpeoneview/cmdlets/v5.40/servers/show-ovportstatistics) Cmdlet is for.  Although that Cmdlet is in the wrong category (`servers`, and should be in `networking`).

**Comment by davidvf** [comment]
Hi @ChrisLynchHPE, I am looking for CPU, Memory, Power and Temperature statistics for the IC modules. Those are visible for the synergy interconnects in OneView except for the SAS module.

**Comment by ChrisLynchHPE** [comment]
Sure, I'll look into it.
